### PR TITLE
Smoke tests to be run after a deploy

### PR DESCRIPTION
**Pull request by aldavidson**
Smoke tests are a pre-requisite for handing over to the Incident Response team, and a useful thing to have anyway.

### CODE REVIEW

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        url = "{root}{endpoint}".format(root=self.root_url, endpoint=endpoint)
+
+        if "?" not in url:
+            url = url + '?'
+
+        if self.auth_token:
+            url = "{url}&token={token}".format(url=url, token=self.auth_token)
+
+        return "{url}&format=json".format(url=url)
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: You should use `urlparse.urlunparse` or `urlparse.urlunsplit` for this.

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import requests
+
+
+class GenuineApi(object):
+
+    def __init__(self, *args, **kwargs):
+        self.root_url = kwargs.get('root_url')
+        self.auth_token = kwargs.get('auth_token')
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: This should be:
```python
def __init__(self, root_url=None, auth_token=None):
    self.root_url = root_url
    self.auth_token = auth_token
```

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            url = url + '?'
+
+        if self.auth_token:
+            url = "{url}&token={token}".format(url=url, token=self.auth_token)
+
+        return "{url}&format=json".format(url=url)
+
+    def get(self, endpoint):
+        u = self._full_url(endpoint)
+        print 'hitting ' + u
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: Remove `print` lines or change to `logging.debug`

Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: Also, `u` is too short a variable name

File: postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from django.test.runner import DiscoverRunner
+
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py]: Missing blank line

File: postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+class NoDbTestRunner(DiscoverRunner):
+
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py]: Extra blank line should be removed

File: postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import os
+
+from django.test import TestCase
+
+from postcode_api.models import Address
+from postcode_api.smoke_tests.genuine_api import GenuineApi
+
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py]: Missing blank line

File: postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    def _make_request(self, endpoint):
+        r = self.api.get(endpoint)
+        self.assertEqual(200, r.status_code)
+        return r.json()
+
+    def test_that_there_is_at_least_one_address(self):
+        r = self._make_request('/addresses/?postcode=n28as')
+        self.assertGreater(0, len(r))
+        self.assertEqual( True, r[0].items().issubset(dir(Address())) )
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py]: This is fine, but I would use `assertTrue` here

File: postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+    def test_that_there_is_a_local_authority(self):
+        r = self._make_request('/postcodes/n28as')
+        assert r['local_authority'] is not None
+        assert r['local_authority']['name']
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py]: Again, no problem - just an odd departure from consistency

File: postcodeinfo/apps/postcode_api/smoke_tests/test_monitoring.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    def _make_request(self, endpoint):
+        r = self.api.get(endpoint)
+        self.assertEqual(200, r.status_code)
+        return r.json()
+
+    def test_that_ping_dot_json_is_ok(self):
+        self._make_request('/ping.json')
+
+    def test_that_healthcheck_dot_json_is_ok(self):
+        self._make_request('/healthcheck.json')
```
Comment (andyhd) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_monitoring.py]: As these two tests make the same method call, maybe they could be data driven. Eg:
```python
urls = [
    '/ping.json',
    '/healthcheck.json']
for url in urls:
    self._make_request(url)
```
Also, I think `_make_request` would be better named `assertHttp200Response` or similar.

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py (hunk 1 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: rewritten to use urlparse methods

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py (hunk 2 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: yup, done

File: postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py (hunk 3 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/genuine_api.py]: done

File: postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py (hunk 4 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py]: fixed

File: postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py (hunk 5 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/no_db_test_runner.py]: fixed

File: postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py (hunk 6 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py]: fixed

File: postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py (hunk 7 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_api_responses.py]: coming back to this with a fresher head, it's actually not a good test, as an empty set would still be a valid subset. So I've removed this test pending some more thought (it's still testing that the array is not empty)

File: postcodeinfo/apps/postcode_api/smoke_tests/test_monitoring.py (hunk 9 above)
Comment (aldavidson) [inline postcodeinfo/apps/postcode_api/smoke_tests/test_monitoring.py]: good point - i've renamed it

**Comment by andyhd** [comment]
:+1:
